File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.Buffer;
import java.util.StringTokenizer;

public class Main {
  static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
  static StringTokenizer st;
  private static int K, N;
  private static int[] arr;
  private static int max;
  private static long answer;

  public static void main(String[] args) throws IOException {
    st = new StringTokenizer(br.readLine());
    K = Integer.parseInt(st.nextToken());
    N = Integer.parseInt(st.nextToken());

    arr = new int[K];
    for (int i = 0; i < K; i++) {
      int ele = Integer.parseInt(br.readLine());
      arr[i] = ele;
      max = Math.max(max, ele);
    } // 입력완료

    binarySearch(N, 1, max);

    System.out.println(answer);
  } // end of main

  private static void binarySearch(int targetN, long start, long end) {
    while (start <= end) {
      long mid = (((long) start + end) / 2);
      int result = 0;
      // 몇개의 랜선이 나오는지 구현해보기
      for (int i = 0; i < K; i++) {
        result += arr[i] / mid;
      }
      // 랜선이 목표치보다 많거나 같게 나오면 랜선의 길이를 올려서 다시해보기
      if (result >= targetN) {
        answer = mid;
        start = mid + 1;
      } else if (result < targetN) { // 랜선이 목표치보다 적다면 랜선의 길이를 줄여보기
        end = mid - 1;
      }
    }
  }
} // end of class

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
"""
답의 형태가 연속적인 매개변수 탐색이다
따라서 이분탐색의 형태로 변경할 수 있다
"""

import sys

input = lambda: sys.stdin.readline().strip()

K, N = map(int, input().split())

lst = []
for _ in range(K):
    lst.append(int(input()))

answer = 0

start = 1
end = max(lst)
while start <= end:
    result = 0
    mid = (start + end) // 2
    for ele in lst:
        result += ele // mid
    if result >= N:
        answer = max(answer, mid)
        start = mid + 1
    else:
        end = mid - 1

print(answer)